apps/src/code-studio/makerlab/setupPage.js

```diff
@@ -0,0 +1,207 @@
+          <SetupStep stepStatus={this.state['status-board-components']}
+                     stepId="board-components"
+                     stepName="Board components usable"/>
+        </div>
+    );
+  },
+
+  hide(selector) {
+    this.setState({[`status-${selector}`]: HIDDEN});
+  },
+
+  fail(selector) {
+    this.setState({[`status-${selector}`]: FAILED});
+  },
+
+  spin(selector) {
+    this.setState({[`status-${selector}`]: ATTEMPTING});
+  },
+
+  succeed(selector) {
+    this.setState({[`status-${selector}`]: SUCCEEDED});
+  },
+
+  thumb(selector) {
+    this.setState({[`status-${selector}`]: CELEBRATING});
+  },
+
+  componentDidMount() {
+    const isChrome = !!window.chrome;
+    const gtChrome33 = isChrome && getChromeVersion() >= 33;
+
+    function getChromeVersion() {
+      const raw = navigator.userAgent.match(/Chrom(e|ium)\/([0-9]+)\./);
+      return raw ? parseInt(raw[2], 10) : false;
+    }
+
+    function isWindows() {
+      return navigator.platform.indexOf('Win') > -1;
+    }
+
+    if (isChrome) {
+      if (gtChrome33) {
+        this.succeed('is-chrome');
+      } else {
+        $('is-chrome').append("Your Chrome version is " + getChromeVersion() + ", please upgrade to at least version 33");
+      }
+
+      const bc = new BoardController();
+      Promise.resolve().then(() => {
+        this.spin('app-installed');
+        return bc.ensureAppInstalled()
+            .then(() => this.succeed('app-installed'))
+            .catch((error) => this.fail('app-installed'));
+      }).then(() => {
+        this.spin('board-plug');
+        return BoardController.getDevicePort()
+            .then(() => this.succeed('board-plug'))
+            .catch((error) => this.fail('board-plug'));
+      }).then(() => {
+        this.spin('board-connect');
+        return bc.ensureBoardConnected()
+            .then(() => this.succeed('board-connect'))
+            .catch((error) => this.fail('board-connect'));
+      }).then(() => {
+        this.spin('board-components');
+        return bc.connectWithComponents()
+            .then(() => this.thumb('board-components'))
+            .then(() => {
+              bc.prewiredComponents.buzzer.play({
+                song: [
+                  ["G3", 100], ["C4", 100], ["E4", 100], ["G4", 50],
+                  [null, 150], ["E4", 75], ["G4", 400]
+                ], tempo: 45000
+              });
+              bc.prewiredComponents.colorLeds.forEach(l => l.color('green'));
+            })
+            .then(() => promiseWaitFor(1600))
+            .then(() => {
+              bc.prewiredComponents.colorLeds.forEach(l => l.off());
+              bc.prewiredComponents.buzzer.play({
+                song: [["C4", 100]],
+                tempo: 45000
+              });
+            })
+            .then(() => this.succeed('board-components'))
+            .catch((error) => this.fail('board-components'));
+      });
+    } else {
+      this.fail('is-chrome');
```

Comment:
This is effectively an early-out; maybe should move it to the top and flatten your logic.